modules/redcap/php/configurations/redcapconfiguration.class.inc

```diff
@@ -0,0 +1,239 @@
+<?php declare(strict_types=1);
+
+/**
+ * PHP Version 8
+ *
+ * @category REDCap
+ * @package  Main
+ * @author   Regis Ongaro-Carcy <regis.ongaro-carcy@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+
+namespace LORIS\redcap\configurations;
+
+
+/**
+ * This represents a REDCap configuration.
+ *
+ * @category REDCap
+ * @package  Main
+ * @author   Regis Ongaro-Carcy <regis.ongaro-carcy@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+class RedcapConfiguration
+{
+    /**
+     * A LORIS User object.
+     *
+     * @var \User
+     */
+    private \User $_assignee;
+
+    /**
+     * A list of REDCap instances.
+     *
+     * @var RedcapConfigurationInstance[]
+     */
+    private array $_instances;
+
+    /**
+     * Constructor.
+     *
+     * @param \User $user a LORIS User
+     */
+    public function __construct(\User $user)
+    {
+        $this->_assignee  = $user;
+        $this->_instances = [];
+    }
+
+    /**
+     * Get the main assignee for all issues.
+     *
+     * @return \User
+     */
+    public function getAssignee(): \User
+    {
+        return $this->_assignee;
+    }
+
+    /**
+     * Add a new REDCap instance.
+     *
+     * @param string $name the instance name
+     * @param string $url  the instance url
+     *
+     * @return void
+     */
+    public function addInstance(string $name, string $url): void
+    {
+        if ($this->hasInstanceName($name)) {
+            throw new \LorisException(
+                "[redcap][config] instance '$name' already exists."
+            );
+        }
+        $this->_instances[$name] = new RedcapConfigurationInstance($name, $url);
+    }
+
+    /**
+     * Get a REDCap instance by name.
+     *
+     * @param string $instanceName the instance name
+     *
+     * @return ?RedcapConfigurationInstance a REDCap instance or null.
+     */
+    public function getInstanceByName(
+        string $instanceName
+    ): ?RedcapConfigurationInstance {
+        return $this->_instances[$instanceName] ?? null;
+    }
+
+    /**
+     * Get a REDCap instance by URL.
+     *
+     * @param string $instanceURL the instance URL
+     *
+     * @return ?RedcapConfigurationInstance a REDCap instance or null.
+     */
+    public function getInstanceByURL(
+        string $instanceURL
+    ): ?RedcapConfigurationInstance {
+        // normalize the given URL
+        $searchedURL = self::_normalizeURL($instanceURL);
+        $i           = null;
+        foreach ($this->_instances as $instance) {
+            // normalize instance URL
+            $iURL = self::_normalizeURL($instance->getURL());
+            // compare
+            if ($iURL === $searchedURL) {
+                $i = $instance;
+                break;
+            }
+        }
+        return $i;
+    }
```

Comment:
This function could be simplified if the instance URL always has a "normalized" URL, and by using `array_find` instead of a `for` loop.

Example:
```php
    public function getInstanceByURL(
        string $instanceURL
    ): ?RedcapConfigurationInstance {
        // normalize the given URL
        $searchedURL = removeTrailingSlash($instanceURL);
        return array_find(
            $this->_instances,
            fn($instance) => $instance->getURL() == $searchedURL,
        );
    }
```